File: vra8-m-ruby-edition/voice.rb
require_relative 'common'
require_relative 'vco'
require_relative 'vcf'
require_relative 'vca'
require_relative 'eg'
require_relative 'lfo'
require_relative 'slew-rate-limiter'

class Voice
  def initialize
    @vco = VCO.new
    @vcf = VCF.new
    @vca = VCA.new
    @eg  = EG.new
    @lfo = LFO.new
    @srl = SlewRateLimiter.new
    @note_number = NOTE_NUMBER_MIN
  end

  def note_on(note_number)
    if ((note_number >= NOTE_NUMBER_MIN) && (note_number <= NOTE_NUMBER_MAX))
      @note_number = note_number
      @eg.note_on
    end
  end

  def note_off
    @eg.note_off
  end

  def control_change(controller_number, controller_value)
    case (controller_number)
    when LFO_RATE
      @lfo.set_rate(controller_value)
    when LFO_RATE_EG_AMT
      # todo
    when VCO_MIX
      @vco.set_pulse_saw_mix(controller_value)
    when VCO_MIX_EG_AMT
      # todo
    when VCO_PULSE_WIDTH
      @vco.set_pulse_width(controller_value)
    when VCO_SAW_SHIFT
      @vco.set_saw_shift(controller_value)
    when VCO_COLOR_EG_AMT:
      # todo
    when VCO_COLOR_LFO_AMT
      @vco.set_color_lfo_amt(controller_value)
    when VCF_CUTOFF
      @vcf.set_cutoff(controller_value)
    when VCF_RESONANCE
      @vcf.set_resonance(controller_value)
    when VCF_CUTOFF_EG_AMT
      @vcf.set_cv_amt(controller_value)
    when VCA_GAIN
      @vca.set_gain(controller_value)
    when EG_ATTACK
      @eg.set_attack(controller_value)
    when EG_DECAY_RELEASE
      @eg.set_decay_release(controller_value)
    when EG_SUSTAIN
      @eg.set_sustain(controller_value)
    when PORTAMENTO
      @srl.set_slew_time(controller_value)
    when ALL_NOTES_OFF
      @eg.note_off
    end
  end

  def clock
    eg_output = @eg.clock
    lfo_output = @lfo.clock
    srl_output = @srl.clock(@note_number << 8)
    vco_output = @vco.clock(srl_output, lfo_output)
    vcf_output = @vcf.clock(vco_output, eg_output)
    vca_output = @vca.clock(vcf_output, eg_output)
    return high_sbyte(vca_output)
  end
end

File: vra8-m-html-edition/voice.js
var Voice = function() {
  this.initialize = function() {
    this.vco = new VCO();
    this.vcf = new VCF();
    this.vca = new VCA();
    this.eg  = new EG();
    this.lfo = new LFO();
    this.srl = new SlewRateLimiter();
    this.note_number = NOTE_NUMBER_MIN;
  };

  this.note_on = function(note_number) {
    this.note_number = note_number;
    this.eg.note_on();
  };

  this.note_off = function() {
    this.eg.note_off();
  };

  this.control_change = function(controller_number, controller_value) {
/*
    switch (controller_number) {
    case VCO_PULSE_SAW_MIX:
      this.vco.set_pulse_saw_mix(controller_value);
      break;
    case VCO_PULSE_WIDTH:
      this.vco.set_pulse_width(controller_value);
      break;
    case VCO_SAW_SHIFT:
      this.vco.set_saw_shift(controller_value);
      break;
    case VCF_CUTOFF:
      this.vcf.set_cutoff(controller_value);
      break;
    case VCF_RESONANCE:
      this.vcf.set_resonance(controller_value);
      break;
    case VCF_EG_AMT:
      this.vcf.set_cv_amt(controller_value);
      break;
    case VCA_GAIN:
      this.vca.set_gain(controller_value);
      break;
    case EG_ATTACK:
      this.eg.set_attack(controller_value);
      break;
    case EG_DECAY_RELEASE:
      this.eg.set_decay_release(controller_value);
      break;
    case EG_SUSTAIN:
      this.eg.set_sustain(controller_value);
      break;
    case LFO_RATE:
      this.lfo.set_rate(controller_value);
      break;
    case LFO_VCO_COLOR_AMT:
      this.vco.set_color_lfo_amt(controller_value);
      break;
    case PORTAMENTO:
      this.srl.set_slew_time(controller_value);
      break;
    case ALL_NOTES_OFF:
      this.eg.note_off();
      break;
    }
*/
  };

  this.clock = function() {
/*
    var eg_output = this.eg.clock;
    var lfo_output = this.lfo.clock;
    var srl_output = this.srl.clock(this.note_number << 8);
    var vco_output = this.vco.clock(srl_output, lfo_output);
    var vcf_output = this.vcf.clock(vco_output, eg_output);
    var vca_output = this.vca.clock(vcf_output, eg_output);
    return high_sbyte(vca_output);
*/
    return this.vco.clock(this.note_number << 8, 0);
  };

  this.initialize();
};
